File: FastHCF.cs
using System;
using System.Collections.Generic;

namespace Arithmetic_CS
{
    class FastHCF
    {
        private int[] set_of_numbers;
        private int[] arg_copy; // Java passes arrays by reference; make a copy.
        private List<int> common_factors = new List<int>(); // factors common to our set_of_numbers
        private List<int> min_prime_factors = new List<int>();

        private int index; // index into array common_factors
        private bool all_round_factor; // intiable to keep state
        private int calc_result;

        public FastHCF(List<int> group)
        {
            set_of_numbers = new int[group.Count];
            arg_copy = new int[group.Count];
            index = 0;

            // STEP 1:
            group.Sort();
            //iterate through and retrieve members
            foreach (int number in group)
            {
                set_of_numbers[index] = number;
                arg_copy[index] = number;
                index++;
            }

            all_round_factor = false;
            calc_result = 1;
        }

        // STEP 2:
        /*
         * Compiles only the factors of the smallest member of 'group_of_numbers'
         */
        private int onlyPrimeFactors(int in_question)
        {
            int temp_limit;
            temp_limit = (int)Math.Ceiling(Math.Sqrt(in_question));

            while (index <= temp_limit)
            {
                if ((in_question % index) == 0)
                {
                    min_prime_factors.Add(index);
                    return onlyPrimeFactors(in_question / index);
                }
                index++;
            }
            min_prime_factors.Add(in_question);

            return 0;
        }

        /**
         * This function searches for mutual factors using an already computed list
         * of factors(for the smallest member of 'set_of_numbers').
         *
         * @return - Nil
         */
        private int findHCFFactors()
        {
            while (index < min_prime_factors.Count)
            {
                all_round_factor = true;
                // STEP 3:
                for (int j = 0; j < arg_copy.Length; j++)
                {
                    if (!(all_round_factor == true
                            && (arg_copy[j] % min_prime_factors[index]) == 0))
                    {
                        all_round_factor = false;
                    }
                }

                // STEP 4:
                if (all_round_factor == true)
                {
                    for (int j = 0; j < arg_copy.Length; j++)
                    {
                        arg_copy[j] /= min_prime_factors[index];
                    }
                    common_factors.Add(min_prime_factors[index]);
                }
                index++;
            }

            return 0;
        }

        /**
         * Just calls out and collects the prepared factors.
         *
         * @return - String value
         */
        public int getHCF()
        {
            index = 2;
            onlyPrimeFactors(set_of_numbers[0]);

            index = 0;
            findHCFFactors();

            //iterate through and retrieve members
            foreach (int factor in common_factors)
            {
                calc_result *= factor;
            }

            return calc_result;
        }
    }
}

File: FastHCF.js
var common_factors = []; // factors common to our group_of_numbers
var index = 0; // index into array common_factors
var state_all_round_factor = false; // variable to keep state
var found_prime_factors = [];
var count = 0; // index into array 'found_prime_factors'

/* Pass of Array(according to javascript) is by reference;
 so make a copy that will be passed instead. */
var arg_copy = [];

var i;

// STEP 2:
/*
 Compiles only the factors of the smallest member of 'group_of_numbers'
 */
function onlyPrimeFactors(sub_level) {
    var temp_limit = Math.ceil(Math.sqrt(sub_level));

    while (i <= temp_limit) {
        if (i != 1 && (sub_level % i) == 0) { // avoid an infinte loop with the i != 1 check.
            found_prime_factors[count++] = i;
            return onlyPrimeFactors(sub_level / i);
        }
        i++;
    }
    found_prime_factors[count] = sub_level;

    return;
}

/*
 * Our function checks 'group_of_numbers';
 * If it finds a factor common to all fo it, it records this factor;
 * then divides 'group_of_numbers' by the common factor found
 * and makes this the new 'group_of_numbers'.
 * It continues recursively until all common factors are found.
 */
function getHCFFactors(operate_on) {
    while (i < found_prime_factors.length) {
        state_all_round_factor = true;
        // STEP 3:
        for (var j = 0; j < operate_on.length; j++) {
            if (state_all_round_factor == true && (operate_on[j] % found_prime_factors[i]) == 0) {
                state_all_round_factor = true;
            } else {
                state_all_round_factor = false;
            }
        }
        // STEP 4:
        if (state_all_round_factor == true) {
            for (var j = 0; j < operate_on.length; j++) {
                operate_on[j] /= found_prime_factors[i];
            }
            common_factors[index++] = found_prime_factors[i];
        }
        i++;
    }
    return;
}

function getFastHCF(group) {
    // STEP 1:
    // Sort in ascending order
    group_of_numbers.sort(
            function (a, b) {
                return a - b;
            }
    );
    i = 2; // start checking for factors from variable 2
    onlyPrimeFactors(group[0]);

    for (var j = 0; j < group.length; j++) {
        arg_copy[j] = group[j];
    }
    i = 0;
    getHCFFactors(arg_copy); // call to our function

    var HCF = 1;
    for (var k = 0; k < common_factors.length; k++) {
        HCF *= common_factors[k];
    }

    return HCF;
}